### PR TITLE
[Tech spec] Inactive user task re-assignment

### PR DESCRIPTION
**Pull request by lowellrex**
Connects #11811.

## Work remaining to be done:
* [ ] Determine which approach(es) we want to pursue
* [ ] Write tickets to do the implement the chosen approach(es)

### CODE REVIEW

File: docs/11811_tech_spec.md [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    - Expect that determination to be made by the product team in consultation with Board employees, as described in [this Slack post by LP](https://dsva.slack.com/files/T03FECE8V/FN5GLN4ES?origin_team=T03FECE8V) (reproduced as a standalone document in this PR)
+
+## Assumptions
+
+* This document assumes that this process begins after the application has identified the inactive user (or user who just left an organization) whose open tasks we want to handle.
+* The existing reassign() function will satisfy the mechanical problem of reassigning tasks even when they are not the terminal node in a branch of the task tree. That is, if a user becomes inactive, tasks that are children of tasks assigned to the newly-inactive user will still be available for action.
+* The application does not care if the task's assignee is no longer a member of the organizationally-assigned parent task is assigned when we are re-assigning their tasks. That is, if we are re-assigning the inactive user Billie's `AojColocatedTask` which has a parent `AojColocatedTask` assigned to the `Colocated` organization, we do not care if Billie is still a member of the `Colocated` organization or not.
```
Comment (yoomlam) [inline docs/11811_tech_spec.md]: Line 22: Hard to parse that first sentence. In the second sentence, should the child and parent tasks be the same?

File: docs/11811_tech_spec.md [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+     2 | PendingScanningVbmsColocatedTask                 | PendingScanningVbmsColocatedTask                 | Organization     | VSCLHALL
+     2 | ScheduleHearingColocatedTask                     | ScheduleHearingColocatedTask                     | Organization     | VSCLHALL
+     1 | PowerOfAttorneyRelatedMailTask                   | PowerOfAttorneyRelatedMailTask                   | Organization     | VSCLHALL
+     1 | TimedHoldTask                                    | OtherColocatedTask                               | User             | VSCLHALL
+     1 | TimedHoldTask                                    | StayedAppealColocatedTask                        | User             | VSCLHALL
+(37 rows)
+```
+
+## Footnotes
+
```
Comment (yoomlam) [inline docs/11811_tech_spec.md]: Are these task relationships enforced in the code to prevent future code changes from breaking these constraints?

File: docs/11811_tech_spec.md [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    if task.parent.assigned_to.is_a?(Organization)
+      task.parent.update!(status: Constants.TASK_STATUSES.assigned)
+      task.update!(status: Constants.TASK_STATUSES.cancelled)
+    end
+
+    ...
+  end
+end
+```
+
```
Comment (yoomlam) [inline docs/11811_tech_spec.md]: Without an understanding of task states, it doesn't make intuitive sense for the parent task to become `assigned`. What was the state before that? Perhaps `needsReview` or `subtasksChanged` conveys the need for reassignment of child tasks.

File: docs/11811_tech_spec.md [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+      # Update the instructions with a note detailing what we've done.
+      task.update!(instructions: [task.instructions, "Re-assigning open task for inactive user"].flatten)
+    end
+  end
+end
+```
+
+## Additional considerations
+
+Additionally, it may be worthwhile to alert folks via email (or some means other than simply task re-assignment which relies on somebody checking a particular queue). Electing to send an email in addition to or in place of re-assigning a task does not change this implementation in any meaningful way since we will still need to identify the recipient of such an email (a `Organization`'s administrator, for instance) and will rely on the task tree to identify that recipient.
```
Comment (yoomlam) [inline docs/11811_tech_spec.md]: Any automated reassignments should trigger a notification to someone, depending on the impact of the reassignment (i.e., based on task type probably). That way, if the recipient wants to change the reassignment, they can.

**Comment by hschallhorn** [comment]
Some thoughts that have spun out of actually doing some of the reassignments [here](https://github.com/department-of-veterans-affairs/caseflow/issues/12173#issuecomment-539197954):
1. next_assignee is S L O O O W W W. We opted to somewhat batch the reassignments by only fetching next_assignee once per assignee, but a batch reassignment function may be necessary
2. The reassign function will want a [`current_user`](https://github.com/department-of-veterans-affairs/caseflow/blob/5fc7d88144b0003e17daed97d1cacfd5977973e0/app/models/task.rb#L135) passed to it to set the [`assigned_by_id`](https://github.com/department-of-veterans-affairs/caseflow/blob/5fc7d88144b0003e17daed97d1cacfd5977973e0/app/models/task.rb#L470) of the new task, so that should be passed in as well rather than nil.

**Comment by hschallhorn** [comment]
When talking to design on #12312, it was mentioned that it may be nice for there to be a delay before the actual reassignment happens. Maybe we can put this in a job to give users some time to undo the status change if need be

File: docs/11811_tech_spec.md (hunk 1 above)
Comment (lowellrex) [reply docs/11811_tech_spec.md]: Great catch. Fixing now.

File: docs/11811_tech_spec.md (hunk 3 above)
Comment (lowellrex) [reply docs/11811_tech_spec.md]: Definitely agree that this particular scenario needs refinement. We are relying on somebody (a team administrator perhaps) monitoring the team's queue to notice that a new case has appeared in that queue. Assuming that somebody is monitoring a team queue may prove to be a bad assumption and we may need a better way to notify the team administrator that an appeal has popped into their queue.

I think it may be beyond the scope of this ticket, but perhaps we can add a [data integrity check](https://github.com/department-of-veterans-affairs/caseflow/blob/0170d3e7771d942d2ce788e02e62b20415b0a592/app/jobs/data_integrity_checks_job.rb#L3) that alerts us (or team admins?) when an appeal has not changed status in some number of days (appeals with `JudgeAssignTask`s that have not yet been dispatched, where the most recent `updated_at` value for that appeal's tasks is some number of days ago)?

Comment (lowellrex) [reply docs/11811_tech_spec.md]: Ticketed: https://github.com/department-of-veterans-affairs/caseflow/issues/12395

**Comment by lowellrex** [comment]
Created #12394 as implementation ticket for this tech spec so I'm closing it out.
